client/blocks/login/index.jsx

```diff
@@ -170,12 +204,25 @@ class Login extends Component {
 			translate,
 			twoStepNonce,
 			fromSite,
+			currentUser,
 		} = this.props;
 
+		const { validatedRedirectToQueryParam } = this.state;
+
 		let headerText = translate( 'Log in to your account' );
 		let preHeader = null;
 		let postHeader = null;
 
+		if ( currentUser ) {
+			// Render ContinueAsUser straight away, even before validation.
+			// This helps avoid jarring layout shifts. It's not ideal that the link URL changes transparently
+			// like that, but it is better than the alternative, and in practice it should happen quicker than
+			// the user can notice.
+			postHeader = (
+				<ContinueAsUser user={ currentUser } redirectUrl={ validatedRedirectToQueryParam || '/' } />
```

Comment:
`ContinueAsUser` could retrieve the `currentUser` from Redux itself (it's a block after all). And would be also a nicer place for the validation logic. It could also be in some "disabled but rendered" state while validation is in progress. In the end, we could just render:
```jsx
<ContinueAsUser />
```
and the component does all the heavy lifting internally.

Reply:
That's a very good point, and I should have thought of that 😞 

It's done! I opted for leaving `currentUser` in the login block, though, as it makes more sense, semantically speaking, to check for a current user before adding a `ContinueAsUser` block to the output. Let me know if you disagree.